text/0281-es5-getters.md

```diff
@@ -0,0 +1,180 @@
+- Start Date: 2017-12-12
+- RFC PR: https://github.com/emberjs/rfcs/pull/281
+- Ember Issue: (leave this empty)
+
+# Summary
+
+Install ES5 getters for computed properties on object prototypes, thus
+eliminating the need to use `this.get()` or `Ember.get()` to access them.
+
+Before:
+
+```js
+import Object, { computed } from '@ember/object';
+
+const Person = Object.extend({
+  fullName: computed('firstName', 'lastName', function() {
+    return `${this.get('firstName')} ${this.get('lastName')}`;
+  })
+});
+
+let chancancode = Person.create({ firstName: 'Godfrey', lastName: 'Chan' });
+
+chancancode.get('fullName'); // => 'Godfrey Chan'
+
+chancancode.set('firstName', 'ʎǝɹɟpo⅁');
+
+chancancode.get('fullName'); // => 'ʎǝɹɟpo⅁ Chan'
+
+let { firstName, lastName, fullName } = chancancode.getProperties('firstName', 'lastName', 'fullName');
+```
+
+After:
+
+```js
+import Object, { computed } from "@ember/object";
+
+const Person = Object.extend({
+  fullName: computed('firstName', 'lastName', function() {
+    return `${this.firstName} ${this.lastName}`;
+  })
+});
+
+let chancancode = Person.create({ firstName: 'Godfrey', lastName: 'Chan' });
+
+chancancode.fullName; // => 'Godfrey Chan'
+
+chancancode.set('firstName', 'ʎǝɹɟpo⅁');
+
+chancancode.fullName; // => 'ʎǝɹɟpo⅁ Chan'
+
+let { firstName, lastName, fullName } = chancancode; // No problem!
+```
+
+# Motivation
+
+Ember inherited its computed properties functionality from [SproutCore](http://guides.sproutcore.com/core_concepts_kvo.html).
+The feature was designed at a time before [ES5 getters](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/get)
+were widely available. This necessitated using a special function such as
+`this.get()` or `Ember.get()` to access the values of computed properties.
+
+Since all of our [target browsers](https://github.com/emberjs/rfcs/pull/252)
+support ES5 getters now, we can drop the need of this special function,
+improving developer egornomics and interoperability between other libraries
+and tooling (such as TypeScript).
```

Comment:
I'm not sure it should be said here. But we can improve ergonomics of tooling such as TypeScript today with `Ember.get`. And if we will able to pass an Array to `Ember.get` like `Ember.get(["some", "nested", "key"])` we would even have better ergonomics with TS

Reply:
I have tried in the past, and I am pretty convinced you cannot type an Ember object successfully today, even without considering nesting. You need to convince TypeScript that they are no properties on the object itself (so obj.foo etc are type errors) but when you do `obj.get(“foo”)` or `get(obj, “foo”)` it should have the right type. I have come close with using some phantom associated type tricks but I think it’s fundamentally impossible.

Reply:
For what it’s worth, you can, and we are... as long as you don’t do nested properties. It took a fairly ridiculous amount of work, though.

Reply:
To be precise: the types you end up with on the object itself end up having a type of `ComputedProperty<TheType>`, which the type system then properly unwraps when using `get`.